### ISSUE TITLE
Go code in need of some major refactoring

### ISSUE THREAD

**Issue by simvux** [probonopd/go-appimage#153]
Hello! I skimmed over the code for appimaged when I created the pull request #152. 

I've gotten the impression that you're a C/C++ developer who's learning Go (correct me if I'm wrong). 
The code does at least seem like it's C/C++ code written in Go syntax. There's several key aspects of ordinary Go development missing such as methods and returning errors. 

For example; 

- functions such as [this](https://github.com/probonopd/go-appimage/blob/master/src/appimaged/commands.go#L12) should return possible errors as the `error` interface. Calling `os.Exit` is something you *very rarely* want to do. A program should in most cases exit by calling `return` from main
- large amounts of global variables are highly discouraged since it makes code less reusable as they depend on a global state. Try to instead collect values needed for specific things in a struct and then creating methods for that struct, rather than having a bunch of plain functions fetching global variables. 
- functions/methods of [this size](https://github.com/probonopd/go-appimage/blob/master/src/appimaged/appimaged.go#L258) should definitely be refactored into smaller ones. 

There's some features I'm looking for such as being able to inject special options into desktop entries generated by appimaged, however since this project is written in a way that's fairly disoriented for a Go project, it'd be a bigger hurdle then I'd like. 

*If* i contribute to implement some extra features, how open are you to me also refactoring some bits of code to be more Go-like as i go?

**Comment by probonopd** [comment]
Hello @simvux thanks for your remarks. I will admit freely that this was my first Go project, ever.

> The code does at least seem like it's C/C++ code written in Go syntax. 

More like a bash script written in Go actually.

> functions such as this should return possible errors as the error interface

Correct, and I'd definitely and quickly want to merge PRs that would improve this. If you'd like to send PRs improving this, then please do not mix the same PRs with any other aspects.

> large amounts of global variables are highly discouraged

At the time I wrote the code this seemed the simplest way to achieve what I need. I'd probably merge PRs that improve this _if_ they don't significantly complicate/increase the code. I am looking for the most simple code that gets the job done usually.

> functions/methods of this size could definitely be refactored into smaller ones

_Could_, but I am not a fan of breaking things down just for the sake of breaking things down. To me, linear program parts are the easiest to read because you can just read the code like a cooking recipe without having to jump all over the place. So it'd be a good idea to briefly discuss changes before sending a PR.

**Before** starting any serious work though we should try to get https://github.com/probonopd/go-appimage/pull/104/files merged though if @srevinsaju is still interested, so that we don't run into merge conflicts later on.

Thanks for your suggestions, I'm always happy to improve my use of Go.

**Comment by srevinsaju** [comment]
>> functions/methods of this size could definitely be refactored into smaller ones
>
> Could, but I am not a fan of breaking things down just for the sake of breaking things down. To me, linear program parts are the easiest to read because you can just read the code like a cooking recipe without having to jump all over the place. So it'd be a good idea to briefly discuss changes before sending a PR.

I agree with @simvuk. The code readability of `go-appimage` is very poor because of @probonopd 's desire to keep it linear. #104 aims to slightly split the older code base to slightly smaller functions, but there is scope for further refactoring nevertheless.

> Before starting any serious work though we should try to get https://github.com/probonopd/go-appimage/pull/104/files merged though if @srevinsaju is still interested, so that we don't run into merge conflicts later on.

I have marked it as 'ready for review'.

**Comment by probonopd** [comment]
Thank you very much @srevinsaju.
It is OK for me to break long functions into smaller units if it helps to make things _easier_. Especially with C++ code my experience has been that it is hard to read (other people's) code if it is not written linearly. linuxdeployqt, which started out by inheriting macdeployqt code, is such an example. (Looks like spaghetti to me to this date.) So I'm not categorically against splitting things up but let's do so moderately.

@simvux is this PR going in a direction you like?

**Comment by simvux** [comment]
#104 does indeed seem like an improvement. 

Although I must say I'm rather surprised by our interaction at #152 and I'm not sure if I'll be able to contribute much since our views on how a project should be developed appear to be drastically different.

**Comment by Flashwalker** [comment]
Go code must die.
Soon it will compile telemetry into every program.
This way it will become malware.
Linux users do not like telemetry for the most part.
